WhateverGreen/kern_igfx.cpp

```diff
@@ -651,6 +721,585 @@ int IGFX::wrapReadAUX(void *that, IORegistryEntry *framebuffer, uint32_t address
+		p1 = 1;
+		p2 = 5;
+	} else if (p == 21) {
+		p0 = 7;
+		p1 = 1;
+		p2 = 3;
+	} else if (p == 35) {
+		p0 = 7;
+		p1 = 1;
+		p2 = 5;
+	}
+	
+	context->pdiv = p0;
+	context->qdiv = p1;
+	context->kdiv = p2;
+}
+
+int IGFX::wrapComputeHdmiP0P1P2(void *that, uint32_t pixelClock, void *displayPath, void *parameters)
+{
+	//
+	// Abstract
+	//
+	// ComputeHdmiP0P1P2 is used to compute required parameters to establish the HDMI connection.
+	// Apple's original implementation cannot find appropriate dividers for a higher pixel clock
+	// value like 533.25 MHz (HDMI 2.0 4K @ 60Hz) and then causes an infinite loop in the kernel.
+	//
+	// This reverse engineering research focuses on identifying fields in CRTC parameters by
+	// carefully analyzing Apple's original implementation, and a better implementation that
+	// conforms to Intel Graphics Programmer Reference Manual is provided to fix the issue.
+	//
+	// This is the first stage to try to solve the HDMI 2.0 output issue on Dell XPS 15 9570,
+	// and is now succeeded by the LSPCON driver solution.
+	// LSPCON is used to convert DisplayPort signal to HDMI 2.0 signal.
+	// When the onboard LSPCON chip is running in LS mode, macOS recognizes the HDMI port as
+	// a HDMI port. Consequently, ComputeHdmiP0P1P2() is called by SetupClock() to compute
+	// the parameters for the connection.
+	// In comparison, when the chip is running in PCON mode, macOS recognizes the HDMI port as
+	// a DisplayPort port. As a result, this method is never called by SetupClock().
+	//
+	// This fix is left here as an emergency fallback and for reference purposes,
+	// and is compatible for graphics on Skylake, Kaby Lake and Coffee Lake platforms.
+	// Note that it is still capable of finding appropriate dividers for a 1080p HDMI connection
+	// and is more robust than the original implementation.
+	//
+	// For those who want to have "limited" 2K/4K experience (i.e. 2K@59Hz or 4K@30Hz) with their
+	// HDMI 1.4 port, you might find this fix helpful.
+	//
+	// For those who have a laptop or PC with HDMI 2.0 routed to IGPU and have HDMI output issues,
+	// it is still recommended to enable the LSPCON driver support to have full HDMI 2.0 experience.
+	//
+	// - FireWolf
+	// - 2019.06
+	//
+	
+	DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: Called with pixel clock = %d Hz.", pixelClock);
+	
+	/// All possible dividers
+	static constexpr uint32_t dividers[] = {
+		// Even dividers
+		4,  6,  8, 10, 12, 14, 16, 18, 20,
+		24, 28, 30, 32, 36, 40, 42, 44, 48,
+		52, 54, 56, 60, 64, 66, 68, 70, 72,
+		76, 78, 80, 84, 88, 90, 92, 96, 98,
+		
+		// Odd dividers
+		3, 5, 7, 9, 15, 21, 35
+	};
+	
+	/// All possible central frequency values
+	static constexpr uint64_t centralFrequencies[3] = {8400000000ULL, 9000000000ULL, 9600000000ULL};
+	
+	// Calculate the AFE clock
+	uint64_t afeClock = pixelClock * 5;
+	
+	// Prepare the context for probing P0, P1 and P2
+	struct ProbeContext context;
+	bzero(&context, sizeof(struct ProbeContext));
```

Comment:
Please do not use struct and do not call bzero. Replace it with just ProbeContext context {}; and let the compiler do the job.

Reply:
Ah, sorry about that. I imported the code from my C project, and I forgot to remove it. : )